### PR TITLE
Edit: Show Accept/Reject codelenses at diff block levels

### PR DESCRIPTION
**Pull request by justinmilner1**
Enables block-level accept/reject of diffs using code lens.

https://github.com/user-attachments/assets/640be261-e7f9-442f-8859-8582f85a99bf

## Test plan

Manual testing:
- Accept all
- only accepts
- only rejects
- mix up order of accepts/rejects
- mixtures of accepts/rejects/acceptAlls
- with deletions + additions

### CODE REVIEW

File: vscode/CHANGELOG.md [hunk 1]
```diff
@@ -9,10 +9,12 @@ This is a log of all notable changes to Cody for VS Code. [Unreleased] changes a
 - Chat: Chat has been added back to the VS Code sidebar (after being removed about 6 months ago). By default, new chats open in the sidebar. New chats can still be opened in an editor panel with the `New Chat in Sidebar` command. Currently open chats can be moved from the sidebar into an editor panel and vice versa. Enterprise users are not affected. [pull/4832](https://github.com/sourcegraph/cody/pull/4832)
 - Chat: Chat History, commands, and settings are now accessible through the chat view for Non-Enterprise users. [pull/4900](https://github.com/sourcegraph/cody/pull/4900)
 - Edit: Added support to select the full range of a file for an edit. [pull/4864](https://github.com/sourcegraph/cody/pull/4864)
+- Edit: Block-level functionality to diff code lens
 
 ### Fixed
 
 - Command: The "Ask Cody to Explain" command for explaining terminal output has been removed from the command palette, as it is only callable from the terminal context menu. [pull/4860](https://github.com/sourcegraph/cody/pull/4860)
+- Autocomplete: Fixed an issue where the cached retriever was attempting to open removed files. [pull/4942](https://github.com/sourcegraph/cody/pull/4942)
```
Comment (justinmilner1) [inline vscode/CHANGELOG.md:17]: Not sure why it's saying I'm committing this... I merged with sourcegraph:main

**Comment by dominiccooney** [comment]
This is cool, I love this idea.

We should get design and product input on this: @toolmantim, @aramaraju

We should also think through supporting this in JetBrains, @taylorsperry

File: vscode/CHANGELOG.md (hunk 1 above)
Comment (dominiccooney) [reply vscode/CHANGELOG.md:17]: Use a rebase workflow, if possible. That is `git rebase --onto origin/main` or just `git rebase` if your upstream is `main`...

**Comment by dominiccooney** [comment]
@justinmilner1 can we add some automated tests for this? Failures here are probably related: vscode/test/e2e/command-edit.test.ts

**Comment by justinmilner1** [comment]
@dominiccooney Will do!

**Comment by toolmantim** [comment]
I love this idea. I think we should add it via a user preference and/or feature flag to start, and then it can go towards a path of figuring out if it makes the right default across all languages etc.

**Comment by aramaraju** [comment]
My POV, the problem, "How do I make it easier for user to accept part of my suggestion", being addressed is a worthwhile thing to solve. I'm a little hesitant on this version of the design though. It's a good first step that we should put out to users to play around with as Tim suggested behind a feature flag, but we will need a few tweaks before making it the default for all users. 

To nitpick, here's a scenario:
There may be an instance where I get 4 blocks and I want to reject one block. I will have to manually accept 3 blocks and reject one. The number of keystrokes/cursor movements may tip this towards being cumbersome for the user. This is not an issue that needs a super complex design to solve though ("Accept all but this" could be a rough solution), but we need to think through such scenarios before making this default IMO

**Comment by justinmilner1** [comment]
@aramaraju The 'accept all' lens remains functional at all times while the diff is open - so in the 'I get 4 blocks and I want to reject one block' scenario, you can just reject the one you don't want, then accept all.

**Comment by aramaraju** [comment]
Just bumping this back up, looks like general consensus is that everyone is super excited about it. Let's get this out :) @justinmilner1 let's add tests to address @dominiccooney 's comments :) when you can

**Comment by justinmilner1** [comment]
@aramaraju @dominiccooney I updated relevant existing tests (which was just one e2e test and manual walkthrough) - let me know if new tests should be created or if I missed something!
 
Also just fyi - I started work on the agent and jetbrains changes, but will be going on vacation in a few days - I may finish them before I leave, but if not, will finish them when I return mid August.

**Comment by umpox** [comment]
Hey @justinmilner1, this is awesome and I want to help out here. I have a little bit of time today to look at this so I'll see if there's anything I can do to move this along and get this merged.

I think we can look at Agent/JetBrains changes in a follow up so I'll see if we can isolate this to VS Code for now

**Comment by jdorfman** [comment]
@justinmilner1 we want to hook you up with Cody credits. Please fill this out https://sourcegraph.com/supporting-open-source

=)

**Comment by umpox** [comment]
Did some testing and think this is safe to merge, will get this into VS Code Insiders today for some more testing in the wild:

Diff variants tested:
- Accept/Rejecting insertions
- Accepting/Rejecting deletions
- Accepting/Rejecting replacements
- Accepting/Rejecting All (with multiple changes)

Edit variants tested:
- Edit
- Add
- Fix
- Document
- Test
- Smart Apply

**Review by umpox — APPROVED** [review]

**Comment by justinmilner1** [comment]
Thanks @umpox !

**Comment by aramaraju** [comment]
🙏 Huge thanks @justinmilner1 for this idea and contribution! Everyone we talked to about this, absolutely loves it 🙇

**Comment by jdorfman** [comment]
@justinmilner1 we gave you a [shoutout in the blog post](https://sourcegraph.com/blog/cody-vscode-1-32-0-release#:~:text=Huge%20thank%20you%20to%20Justin%20Milner%20for%20making%20this%20contribution!%20%E2%9C%A8%20%F0%9F%AB%B6) =)

**Pull request merged**